matchers/parse/parse.go

```diff
@@ -0,0 +1,315 @@
+		return nil, fmt.Errorf("%s: %w", err, ErrExpectedEOF)
+	}
+	return nil, nil
+}
+
+func (p *Parser) parseLabelMatcher(l *Lexer) (parseFn, error) {
+	var (
+		err        error
+		tok        Token
+		labelName  string
+		labelValue string
+		ty         labels.MatchType
+	)
+
+	// The next token is the label name. This can either be an ident which
+	// accepts just [a-zA-Z_] or a quoted which accepts all UTF-8 characters
+	// in double quotes.
+	if tok, err = p.expect(l.Scan, TokenIdent, TokenQuoted); err != nil {
+		return nil, fmt.Errorf("%s: %w", err, ErrNoLabelName)
+	}
+	labelName = tok.Value
+
+	// The next token is the operator such as '=', '!=', '=~' and '!~'.
+	if tok, err = p.expect(l.Scan, TokenOperator); err != nil {
+		return nil, fmt.Errorf("%s: %s", err, ErrNoOperator)
+	}
+	if ty, err = matchType(tok.Value); err != nil {
+		panic("Unexpected operator")
+	}
+
+	// The next token is the label value. This too can either be an ident
+	// which accepts just [a-zA-Z_] or a quoted which accepts all UTF-8
+	// characters in double quotes.
+	if tok, err = p.expect(l.Scan, TokenIdent, TokenQuoted); err != nil {
+		return nil, fmt.Errorf("%s: %s", err, ErrNoLabelValue)
+	}
+	if tok.Kind == TokenIdent {
+		labelValue = tok.Value
+	} else {
+		labelValue, err = strconv.Unquote(tok.Value)
+		if err != nil {
+			return nil, fmt.Errorf("%d:%d: %s: invalid input", tok.ColumnStart, tok.ColumnEnd, tok.Value)
+		}
+	}
+
+	m, err := labels.NewMatcher(ty, labelName, labelValue)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create matcher: %s", err)
+	}
+	p.matchers = append(p.matchers, m)
+
+	return p.parseLabelMatcherEnd, nil
+}
+
+func (p *Parser) parseLabelMatcherEnd(l *Lexer) (parseFn, error) {
+	tok, err := p.expect(l.Peek, TokenComma, TokenCloseBrace)
+	if err != nil {
+		// If this is the end of input we still need to check if the optional
+		// open brace has a matching close brace.
+		if errors.Is(err, ErrEOF) {
+			return p.parseCloseParen, nil
+		}
+		return nil, fmt.Errorf("%s: %s", err, "expected a comma or close brace")
+	}
+	if tok.Kind == TokenCloseBrace {
+		return p.parseCloseParen, nil
+	} else if tok.Kind == TokenComma {
+		return p.parseComma, nil
+	} else {
+		panic("unreachable")
+	}
+}
+
+// Parse parses one or more matchers in the input string. It returns an error
+// if the input is invalid.
+func Parse(input string) (labels.Matchers, error) {
+	p := NewParser(input)
+	return p.Parse()
+}
+
+// ParseMatcher parses the matcher in the input string. It returns an error
+// if the input is invalid or contains two or more matchers.
+func ParseMatcher(input string) (*labels.Matcher, error) {
+	if strings.HasPrefix(input, "{") {
+		return nil, errors.New("Individual matchers cannot start and end with braces")
+	}
+	m, err := Parse(input)
+	if err != nil {
+		return nil, err
+	}
+	if len(m) > 1 {
+		return nil, fmt.Errorf("expected 1 matcher, found %d", len(m))
+	}
+	return m[0], nil
+}
```

Comment:
Can we add some tests to this function?

Reply:
Yes. This function is "special". I'd rather not have it, but it addresses the fact that Alertmanager uses two parse functions with different expected behaviours: the first is for parsing zero or more matchers and is used in operations such as [creating silences](https://github.com/prometheus/alertmanager/blob/e1e59925cc7a4a953c7deba7fdb215224ed05fdb/cli/silence_add.go#L95), while the second is for parsing exactly one matcher and is used in operations such as [querying silences](https://github.com/prometheus/alertmanager/blob/e1e59925cc7a4a953c7deba7fdb215224ed05fdb/cli/silence_query.go#L101).

Reply:
I agree - the querying and creating of silences behave that way to support the use case of specifying the alertname as a single argument matcher  e.g. `amtool query AlwaysFiring` or `amtool query Alwaysfiring team=alerting`.

We don't have to solve that now, as you said, but let's add the tests to make sure we have the needed coverage and we can refactor it later.